rfcs/2020-10-29-3736-remap-chain-parsing.md

```diff
@@ -0,0 +1,181 @@
+
+The same can be said for allowing assignment within an `if` statement. The condition will be valid for both a single (`=`) and double (`==`) equals
+in the condition, yet both variations do very different things:
+
+```
+if .foo == true {
+
+}
+```
+
+```
+if .foo = true {
+
+}
+```
+
+This has haunted C programmers for decades.
+
+Also, the assignment still occurs even if the condition doesn't succeed. Side effects such as this can be unexpected for the script writer
+and could cause unexpected bugs to creep in.
+
+
+## Alternatives
+
+### Enforce boolean conditions
+Allowing for non-boolean conditions in the `if` condition is not strictly necessary. A similar result could be achieved by running the regular
+expression twice, one with `match` which returns a boolean if it matches, and then again with `matches` which would extract the matches out of 
+the text.
+
+```coffeescript
+if match(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/) {
+  $match = matches(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/)
+  .method = $match.method
+  .remote_addr = $match.remote_addr
+  .source = "nginx"
+}
+```
+
+However, this has performance implications as the regular expression has to be run twice.
+
+### Use `switch`
+There are some alternative syntaxes that may be worth considering. One possibility that is common in many languages is a `switch` statement.
+
+```coffeescript
+switch {
+ ($match = matches(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/)): {
+    .method = $match.method
+    .remote_addr = $match.remote_addr
+    .source = "nginx"
+  }
+ ($match = matches(.message, /^(?P<remote_addr>[^\s]*).*"(?P<method>[^\s]*).*"$/): {
+    .method = $match.method
+    .remote_addr = $match.remote_addr
+    .source = "haproxy"
+  }
+  default: {
+    .source = "other"
+  }
+}
+```
+
+The disadvantage of this approach is it is an extra syntactical element that needs to be added to the language that would require maintenance and
+make the language harder to learn. The advantage, however, is that we can then potentially avoid having to enable `if` statements to work with non-boolean 
+values. However, I'm not convinced this would be a good idea as having conditions working differently depending on their context could be confusing
+for the user.
+
+
+## Outstanding Questions
+
+- One thing that the `swimlanes` can do that this RFC hasn't catered for is that `swimlanes` can send the event to different sinks depending on a condition. 
+With Remap there is only one possible component, another transform or a sink, that can accept the output of the mapping. Is this something that should 
+be catered for by the Remap language and this RFC?
+
+
+## Plan Of Attack
+
+Incremental steps that execute this change. Generally this is in the form of:
+
+- [ ] Add `else if` to the remap language parser and add multiple branches to the `IfStatement` struct to handle multiple conditions.
```

Comment:
Note that I had support for this lying around in a local branch, I cleaned that up and pushed it to #4814.

_(I didn't update `IfStatement` to handle multiple conditions, instead wrapping each `false_expression` in a new `IfStatement` expression, but the outcome is the same)_